README.md
# coq2hs-cpp-verify
Proof of concept test bench to code compare functions implemented in Haskell and C++.

Please visit the technical [documentation](https://roycrippen.github.io/coq2hs-cpp-verify/).

requires haskell [stack](https://docs.haskellstack.org/en/stable/README/) >= 1.9.8 and (gcc >= 5.0 or clang >= 3.8)  

## Quick Start

build
```bash
stack build
```

run sample application calling both Haskell and C++ functions
```bash
stack run
```

run property based tests and unit tests comparing results from Haskell and C++ functions
```bash
stack test
```

re-build documentation (stored in ./docs)
```bash
stack haddock
```

load application in GHCI and run functions interactively
```bash
stack ghci  --ghci-options='-fobject-code -O0'
> HsLib.isTriple 3 4 5
> True
> InlineCPP.isTriple 3 4 5
> 1
```

load a single file in GHCI, for example InlineCPP
```bash
stack ghci src/InlineCPP.hs --ghci-options='-fobject-code -O0'
> isTriple 3 4 5
> 1
```

## Concept 
Extract a Haskell function (hsFunction) from a formal [Coq](https://coq.inria.fr/) proof that represents the specification of an existing C++ function (cppFunction).  The extracted Haskell function by definition will be verified as correct. This test bench represents an example of how to use [Haskell](https://www.haskell.org/), [inline-c-cpp](http://hackage.haskell.org/package/inline-c-cpp) and the property based testing tool [QuickCheck](http://hackage.haskell.org/package/QuickCheck) to verify that hsFunction and cppFunction produce identical results for the same inputs. 

## Simple Example
### Haskell function
```Haskell
module HsLib ( square )

-- square a number
square :: Num a => a -> a
square x = x * x
```
### C++ function
```cpp
namespace cn {
    // square a number
    int square(int a) {
        return a * a;
    }
}
```
### Haskell function calling cn::square as inline-cpp
```Haskell
module InlineCPP ( square )

-- inline call to c++ 
square :: CInt -> IO CInt
square x = [C.exp| int { cn::square($(int x)) } |]
``` 
### Haskell QuickCheck test
```Haskell
module Main where

import           Control.Monad
import           Foreign.C.Types
import           Test.Hspec
import           Test.QuickCheck
import           Test.QuickCheck.Monadic
import qualified HsLib                         as HS
import qualified InlineCPP                     as CPP

-- quickCheck property comparing the c++ and Haskell results
prop_square :: Property
prop_square = forAll arbitrary $ \n -> monadicIO $ do
    nn <- run $ CPP.square (n :: CInt)
    assert $ nn == HS.square n

-- run test 10,000 times with random integers
main :: IO ()
main =
    hspec
        $ describe "Haskell checking CPP"
        $ it "quickcheck: HS.square a == CPP.square a"
        $ quickCheck (withMaxSuccess 10000 prop_square)
```

### Run Test 
```bash
$ stack test
Haskell checking CPP
+++ OK, passed 10000 tests.
  quickcheck: HS.square a == CPP.square a
```

### Todo
- 

ChangeLog.md
# Changelog for coq2hs-cpp-bench

## Unreleased changes


test/Spec.hs
module Main where

import           Test.Hspec                     ( shouldBe
                                                , describe
                                                , hspec
                                                , it
                                                )
import           Test.QuickCheck                ( Arbitrary
                                                , Gen
                                                , Property(..)
                                                , arbitrary
                                                , choose
                                                , elements
                                                , forAll
                                                , frequency
                                                , listOf
                                                , sized
                                                , suchThat
                                                , quickCheck
                                                , withMaxSuccess
                                                )
import           Test.QuickCheck.Monadic        ( monadicIO
                                                , run
                                                , assert
                                                )
import           Test.QuickCheck.Instances
import           Foreign.C.Types
import           Control.Monad                  ( (>=>) )
import           Data.ByteString                ( ByteString )
import qualified Data.ByteString               as B
import qualified Data.ByteString.Char8         as C
import qualified InlineCPP                     as CPP
import qualified HsLib                         as HS

main :: IO ()
main = hspec $ do
    describe "HS functions" $ do
        it "square a negative integer" $ do
            HS.square (-3) `shouldBe` 9
            HS.square (-0) `shouldBe` 0
            HS.square (-81) `shouldBe` HS.square 81
        it "square positive integer" $ do
            HS.square (HS.square 2) `shouldBe` 16
            HS.square 9 `shouldBe` 81
        it "(3,4,5) is     a triplet" $ HS.isTriple 3 4 5 `shouldBe` True
        it "(3,4,6) is NOT a triplet" $ HS.isTriple 3 4 6 `shouldBe` False
        it "quickcheck: HS.applyXorCipher (HS.applyXorCipher s k) k == s"
            $ quickCheck (withMaxSuccess 10000 prop_hsApplyXorCipher)
    describe "CPP Functions" $ do
        it "sqaure a negative integer" $ do
            CPP.square (-3) >>= \v -> v `shouldBe` 9
            CPP.square (-0) >>= \v -> v `shouldBe` 0
            v1 <- CPP.square 81
            v2 <- CPP.square (-81)
            v1 `shouldBe` v2
        it "square positive integer" $ do
            CPP.square 2 >>= (CPP.square >=> (`shouldBe` 16))
            CPP.square 9 >>= \v -> v `shouldBe` 81
        it "(3,4,5) is     a triplet" $ CPP.isTriple 3 4 5 >>= \v ->
            v `shouldBe` 1
        it "(3,4,6) is not a triplet" $ CPP.isTriple 3 4 6 >>= \v ->
            v `shouldBe` 0
        it "quickcheck: CPP.applyXorCipher (CPP.applyXorCipher s k) k == s"
            $ quickCheck (withMaxSuccess 10000 prop_cppApplyXorCipher)
    describe "Haskell checking CPP, HS.f args == CPP.f args" $ do
        it "quickcheck: HS.square a == CPP.square a" $ quickCheck prop_square
        it "quickcheck: HS.isTriple a b c == CPP.isTriple a b c"
            $ quickCheck prop_pTriples
        it "quickcheck: HS.applyXorCipher s k == CPP.applyXorCipher s k"
            $ quickCheck (withMaxSuccess 10000 prop_applyXorCipher)

flexList :: Arbitrary a => Gen [a]
flexList = sized
    $ \n -> frequency [(1, return []), (n, (:) <$> arbitrary <*> flexList)]

rangeVal :: Gen Int
rangeVal = choose (3, 1000)

rangeVals :: Gen (Int, Int, Int)
rangeVals = do
    v1 <- rangeVal
    v2 <- rangeVal
    v3 <- rangeVal
    frequency [(1, elements HS.triples), (5, return (v1, v2, v3))]

pTripleVals :: Gen (Int, Int, Int)
pTripleVals = do
    (v1, v2, v3) <- rangeVals `suchThat` (\(v1, v2, v3) -> v1 < v2 && v2 < v3)
    return (v1, v2, v3)

prop_pTriples :: Property
-- prop_pTriples = forAll pTripleVals $ \(a, b, c) ->
--     collect (a, b, c) $ monadicIO $ do
prop_pTriples = forAll pTripleVals $ \(a, b, c) -> monadicIO $ do
    let a' = fromIntegral a
        b' = fromIntegral b
        c' = fromIntegral c
    i <- run (CPP.isTriple a' b' c')
    assert (HS.isTriple a b c == (i == 1))

prop_cppApplyXorCipher :: Property
prop_cppApplyXorCipher = forAll arbitrary $ \s0 -> monadicIO $ do
    let key = C.pack "some string key 1234"
    s1 <- run $ CPP.applyXorCipher s0 key
    s2 <- run $ CPP.applyXorCipher s1 key
    assert $ s2 == s0

prop_hsApplyXorCipher :: Property
prop_hsApplyXorCipher = forAll arbitrary $ \s -> monadicIO $ do
    let key = C.pack "some string key 1234"
    assert $ HS.applyXorCipher (HS.applyXorCipher s key) key == s

prop_applyXorCipher :: Property
prop_applyXorCipher = forAll arbitrary $ \s -> monadicIO $ do
    let key = C.pack "some string key 1234"
    s1 <- run $ CPP.applyXorCipher s key
    assert $ s1 == HS.applyXorCipher s key

prop_square :: Property
prop_square = forAll arbitrary $ \n -> monadicIO $ do
    nn <- run $ CPP.square (n :: CInt)
    assert $ nn == HS.square n


app/Main.hs
module Main where

import qualified Data.Vector.Storable          as V
import qualified InlineCPP                     as CPP
import qualified HsLib                         as HS
import           Foreign.C
import           Control.Monad.IO.Class
import           Data.ByteString                ( ByteString )
import qualified Data.ByteString               as B
import qualified Data.ByteString.Char8         as C

main :: IO ()
main = do
    putStrLn "\nInline CPP compared to Haskell\n"

    -- example hs and cpp result compare
    cppSquare   <- CPP.square 3
    isCppTriple <- CPP.isTriple 3 4 5
    putStrLn $ "CPP.square 3: " ++ show cppSquare
    putStrLn $ " HS.square 3: " ++ show (HS.square 3)
    putStrLn $ "CPP.isTriple 3 4 5: " ++ show isCppTriple
    putStrLn $ " HS.isTriple 3 4 5: " ++ show (HS.isTriple 3 4 5)

    -- example sending a vector to cpp
    let xs = [10, 30, 60]
    x <- CPP.sumVec xs
    putStrLn $ "sum " ++ show xs ++ " = " ++ show x

    -- example getting a cpp created list 
    xs <- CPP.rangeList 10
    putStrLn $ "rangeList 10 = " ++ show xs ++ "\n"

    -- validate calling CPP.applyXorCipher with std::cout
    -- interesting way to quickly test the cpp code
    CPP.testApplyXorCipher

    -- call CPP.applyXorCipher twice
    putStrLn ""
    let s1  = C.pack "String to cipher."
        key = C.pack "my secret key"
    s2 <- CPP.applyXorCipher s1 key
    s3 <- CPP.applyXorCipher s2 key
    putStrLn $ "input string   = " ++ show s1
    putStrLn $ "encoded string from cpp = " ++ show s2
    putStrLn $ "decoded string from cpp = " ++ show s3

    -- call HS.applyXorCipher twice
    putStrLn ""
    let s1  = C.pack "String to cipher."
        key = C.pack "my secret key"
        s2  = HS.applyXorCipher s1 key
        s3  = HS.applyXorCipher s2 key
    putStrLn $ "input string   = " ++ show s1
    putStrLn $ "encoded string from hs = " ++ show s2
    putStrLn $ "decoded string from hs = " ++ show s3



src/HsLib.hs
-- | Haskell equivalent functions of the CPP functions to test for correctness.
module HsLib
    ( square
    , isTriple
    , triples
    , applyXorCipher
    )
where

import           Foreign.C.String               ( castCharToCChar
                                                , castCCharToChar
                                                )
import           Data.Bits                      ( xor )
import           Data.ByteString                ( ByteString )
import qualified Data.ByteString               as B
import qualified Data.ByteString.Char8         as C

-- | Square a number.
square :: Num a => a -> a
square x = x * x

-- | Test whether a, b and c form a  Pythagorean triple given that a < b < c.
-- For example,
--
-- >>> isTriple 7 24 25
-- True
-- >>> isTriple 7 24 26
-- False
isTriple :: Int -> Int -> Int -> Bool
isTriple a b c = square a + square b == square c

-- | List of Pythagorean triples for c < 100. Used to ensure that 20% of 
--  the random triple values generated are valid Pythagorean triples.
triples :: [(Int, Int, Int)]
triples =
    [ (3 , 4 , 5)
    , (5 , 12, 13)
    , (8 , 15, 17)
    , (7 , 24, 25)
    , (20, 21, 29)
    , (12, 35, 37)
    , (9 , 40, 41)
    , (28, 45, 53)
    , (11, 60, 61)
    , (33, 56, 65)
    , (16, 63, 65)
    , (48, 55, 73)
    , (36, 77, 85)
    , (13, 84, 85)
    , (39, 80, 89)
    , (65, 72, 97)
    ]

-- | Used to encode or decode a message with XOR and key. For example,
--
-- >>> import qualified Data.ByteString.Char8 as C
-- >>> applyXorCipher (C.pack "message") (C.pack "my key")
-- "\NUL\FSS\CAN\EOT\RS\b"
-- >>>  applyXorCipher (C.pack "\NUL\FSS\CAN\EOT\RS\b") (C.pack "my key")
-- "message"
applyXorCipher :: ByteString -> ByteString -> ByteString
applyXorCipher msg key = B.pack (B.zipWith xor msg keys)
  where
    l    = C.length msg `div` C.length key
    keys = C.concat $ replicate (l + 1) key




cpp/src/main.cpp
#include <iostream>
#include "../include/candidates.hpp"

using namespace std;

int main(int argc, char *argv[]) {
    int a = 3;
    int b = 4;
    int c = 5;
    cout << "Is (" << a << ", " << b << ", " << c << ") a pythagorean triple: "
         << boolalpha << cn::is_pythagorean_triple(a, b, c) << "\n";
    c = 6;
    cout << "Is (" << a << ", " << b << ", " << c << ") a pythagorean triple: "
         << boolalpha << cn::is_pythagorean_triple(a, b, c) << "\n";

    auto cs = "This is the test string...";
    string key = "cipher key 123";
    cout << "key: '" << key << "'\n";
    cout << "input     string: '" << cs << "'\n";
    auto encrypted = cn::applyXorCipher(cs, key);
    const vector<uint8_t> vec(encrypted.begin(), encrypted.end());
    cout << "encrypted string: '" << cn::show(vec) << "'\n";
    auto decrypted = cn::applyXorCipher(encrypted, key);
    cout << "decrypted string: '" << decrypted << "'\n";
    auto orig_str = cn::applyXorCipher(cn::applyXorCipher(cs, key), key);
    cout << "applyXorCipher(applyXorCipher(cs, key), key): '" << orig_str << "'\n";

    return 0;
}


cpp/include/candidates.hpp
#include <vector>

using namespace std;

namespace cn {
    /// square a number
    int square(int a) {
        return a * a;
    }

    /// do a,b, c form a pythagorean triple
    bool is_pythagorean_triple(int a, int b, int c) {
        return square(a) + square(b) == square(c);
    }

    /// XOR Encryption
    string applyXorCipher(string const &cs, string const &key) {
        string res = cs;
        auto mod = key.size() / sizeof(char);
        for (int i = 0; i < cs.size(); i++) {
            auto c = key[i % mod];
            res[i] = cs[i] ^ c;
        }
        return res;
    }

    /// pretty string of a const vector<T>
    template<typename T>
    string show(const vector<T> &vs) {
        string res = "{ ";
        for (int i = 0; i < vs.size() - 1; i++) {
            res += to_string(vs[i]) + ", ";
        }
        res += to_string(vs[vs.size()]) + " }";
        return res;
    }

    /// pretty string of a vector<T>
    template<typename T>
    string show(vector<T> &vs) {
        const vector<T> vs_ = vs;
        return show(vs_);
    }

} // namespace cn

src/InlineCPP.hs
{-# LANGUAGE TemplateHaskell #-}
{-# LANGUAGE QuasiQuotes #-}
{-# LANGUAGE ScopedTypeVariables #-}
-- | Haskell wrapper functions for using CPP directly as inline source code.
module InlineCPP
  ( square
  , isTriple
  , sumVec
  , rangeList
  , testApplyXorCipher
  , applyXorCipher
  )
where

import           Data.Monoid                    ( (<>) )
import           Foreign.C.Types                ( CInt )
import           Foreign.C.String               ( newCString
                                                , peekCStringLen
                                                )
import           Foreign.Marshal.Array          ( peekArray )
import           Foreign.Marshal.Alloc          ( free )
import           Data.ByteString                ( ByteString )
import qualified Data.ByteString               as B
import qualified Data.ByteString.Unsafe        as BU
import qualified Data.ByteString.Char8         as C
import qualified Data.Vector.Storable          as V
import qualified Data.Vector.Storable.Mutable  as VM
import qualified Language.C.Inline.Cpp         as C
import qualified HsLib                         as HS

C.context (C.cppCtx <> C.vecCtx)
C.include "<iostream>"
C.include "<cstring>"
C.include "../cpp/include/candidates.hpp"

-- brittany-disable-next-binding
-- | Inline call to CPP function cn::square.
--  Equivalent to HS function 'HS.square'
square :: CInt -> IO CInt
square x = [C.exp| int { cn::square($(int x)) } |]

-- brittany-disable-next-binding
-- | Inline call to CPP function cn::isTriple. 
--  Equivalent to HS function 'HS.isTriple'.
isTriple :: CInt -> CInt -> CInt -> IO CInt
isTriple a b c =
   [C.exp| int { cn::is_pythagorean_triple($(int a), $(int b), $(int c)) } |]

-- brittany-disable-next-binding
-- | Inline CPP that sums items in a HS list. For example,
--
-- >>> sumVec [10,20,30]
-- 100
sumVec :: [CInt] -> IO CInt
sumVec xs = do
  -- create mutable vector that cpp block can reference
  vec <- V.thaw (V.fromList xs)

  -- cpp code to sum the list 
  n <- [C.block| int {
    int* xs = $vec-ptr:(int *vec);
    int len = $vec-len:vec;

    int sum = 0;
    for (int i = 0; i < $vec-len:vec; i++) {
        sum += xs[i];
    }
    return sum;
  } |]

  -- return the sum
  _ <- V.freeze vec
  return n

-- brittany-disable-next-binding
-- | Inline CPP that creates a HS list containing [0 .. (n - 1)]. For example,
--
-- >>> rangeList 4
-- [0,1,2,3]
rangeList :: Int -> IO [Int]
rangeList _n = do
  -- cast _n from Int to CInt
  let n = fromIntegral _n

  -- build a list in a cpp block
  arrayPtr <- [C.block| int* {
    int n_ = $(int n);
    int* mat;
    mat = new int[n_];
    for (int i = 0; i < n_; i++) {
      mat[i] = i;
    }
    return mat;
  } |]

  -- copy ptr contents to list of Int
  ls <- peekArray _n arrayPtr
  free arrayPtr
  return $ map fromIntegral ls


-- brittany-disable-next-binding
-- | Inline CPP calls cn::applyXorCipher twice to encoded then decode a message.
-- Demonstrates use of cout and successful round trip use of the XOR cipher.
testApplyXorCipher :: IO ()
testApplyXorCipher = [C.exp| void {
    cout << "This is message from cpp std::cout.\n" ;
    auto cs = "This is the test string...";
    string key = "cipher key 123";
    cout << "std::cout -> (cs, key): ('" << cs << "', '" << key << "')" << "'\n";
    auto orig_str = cn::applyXorCipher(cn::applyXorCipher(cs, key), key);
    cout << "std::cout -> applyXorCipher(applyXorCipher(cs, key), key): '" << orig_str << "'\n";
  }|]

-- brittany-disable-next-binding
-- | Inline call to CPP function cn::applyXorCipher. Equivalent to HS function 'HS.applyXorCipher'.
-- Used to encode or decode a message with XOR and key. For example,
--
-- >>> import qualified Data.ByteString.Char8 as C
-- >>> applyXorCipher (C.pack "message") (C.pack "my key")
-- "\NUL\FSS\CAN\EOT\RS\b"
-- >>>  applyXorCipher (C.pack "\NUL\FSS\CAN\EOT\RS\b") (C.pack "my key")
-- "message"
applyXorCipher :: ByteString -> ByteString -> IO ByteString
applyXorCipher msg key =  do
  (res, size) <- BU.unsafeUseAsCStringLen msg $ \(cStrIn, cStrInLength) ->
    BU.unsafeUseAsCString key $ \cStrKey -> do
      let nCInt = fromIntegral cStrInLength :: CInt
      cStrOut <- [C.block| char* {
        // rename inputs to more readable names
        char *cStrIn = $(char *cStrIn);
        char *key = $(char *cStrKey);
        int n = $(int nCInt);

        // copy *char to string (not casting)
        // necessary because string might contain NULL
        std::string s = "";
        for (int i = 0; i < n; i++) {
          s.push_back(cStrIn[i]);
        }

        auto ds = cn::applyXorCipher(s, key);

        // convert decoded std::string to a char* and return it
        char *cstr = new char[n + 1];
        std::memcpy(cstr, ds.data(), n);
        return cstr;
      }|]
      return (cStrOut, cStrInLength)

  B.packCStringLen (res, size)


